admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -237,3 +302,134 @@ async fn process_submit_transaction_request(
         );
     }
 }
+
+/// Validate transaction signature, then via VM, and add it to Mempool if it passes VM check.
+pub(crate) fn submit_transaction_inner<M, V>(
+    upstream_proxy_data: UpstreamProxyData<M, V>,
+    req: SubmitTransactionRequest,
+) -> failure::Result<SubmitTransactionResponse>
+where
+    M: MempoolClientTrait,
+    V: TransactionValidation,
+{
+    // Drop requests first if mempool is full (validator is lagging behind) so not to consume
+    // unnecessary resources.
+    if !can_send_txn_to_mempool(&upstream_proxy_data)? {
+        debug!("Mempool is full");
+        OP_COUNTERS.inc_by("submit_txn.rejected.mempool_full", 1);
+        let mut response = SubmitTransactionResponse::default();
+        let mut status = MempoolAddTransactionStatus::default();
+        status.set_code(MempoolIsFull);
+        status.message = "Mempool is full".to_string();
+        response.status = Some(Status::MempoolStatus(status));
+        return Ok(response);
+    }
+
+    let txn_proto = req.transaction.clone().unwrap_or_else(Default::default);
+
+    let transaction = match SignedTransaction::try_from(txn_proto.clone()) {
+        Ok(t) => t,
+        Err(e) => {
+            security_log(SecurityEvent::InvalidTransactionAC)
+                .error(&e)
+                .data(&txn_proto)
+                .log();
+            let mut response = SubmitTransactionResponse::default();
+            response.status = Some(Status::AcStatus(
+                AdmissionControlStatus::Rejected("submit txn rejected".to_string()).into(),
+            ));
+            OP_COUNTERS.inc_by("submit_txn.rejected.invalid_txn", 1);
+            return Ok(response);
+        }
+    };
+
+    let gas_cost = transaction.max_gas_amount();
+    let validation_status = upstream_proxy_data
+        .vm_validator
+        .validate_transaction(transaction.clone())
+        .wait()
+        .map_err(|e| {
+            security_log(SecurityEvent::InvalidTransactionAC)
+                .error(&e)
+                .data(&transaction)
+                .log();
+            e
+        })?;
+    if let Some(validation_status) = validation_status {
+        let mut response = SubmitTransactionResponse::default();
+        OP_COUNTERS.inc_by("submit_txn.vm_validation.failure", 1);
+        debug!(
+            "txn failed in vm validation, status: {:?}, txn: {:?}",
+            validation_status, transaction
+        );
+        response.status = Some(Status::VmStatus(validation_status.into()));
+        return Ok(response);
+    }
+    let sender = transaction.sender();
+    let account_state = block_on(get_account_state(
```

Comment:
nit: now you are inside runtime, get rid of all `block_on` calls and make function async

Reply:
Makes sense! It's ok to use block_on in the tests though right? Since I can't make them async and I need to await on submit_transaction_to_mempool to test